### ISSUE TITLE
Key constraint errors when force loading configurations

### ISSUE THREAD

**Issue by juliapampus** [FraunhoferISST/DataspaceConnector#850]
As a side note - the used flag is `configuration.force.reload=true` and it resulted in key constraint errors due to already existing ids in the database.
We have not yet tried DSC **v6.5.3**, as the changelog does not contain related updates.

_Originally posted by @pdrd in https://github.com/International-Data-Spaces-Association/DataspaceConnector/issues/849#issuecomment-1017520128_

**Comment by juliapampus** [comment]
@pdrd Could you tell us what IDs have been duplicated?

**Comment by ghost** [comment]
If I build the current main as a Docker image, run it, stop it, set the mentioned flag in the env variables (`CONFIGURATION_FORCE_RELOAD=true`) without any changes to config.json/application.properties, and boot the image again, it runs with no problems for me.

![image](https://user-images.githubusercontent.com/75306992/150374489-37ee43d9-4a39-43ca-89db-7845579ff79a.png)

The error message / log would be helpful here.

**Comment by juliapampus** [comment]
> without any changes to config.json/application.properties

What if you create a new configmodel with new IDs for all the objects? Or change the values. I guess it's only the `config.json` not the `application.properties` that is causing issues.

**Comment by ghost** [comment]
Tested now, no problems detected, same result as before. `application.properties` is relevant for the config, because keyStorePassword, trustStorePassword and keyAlias are part of this file, which are also persisted into the db / reloaded if required.

Maybe there is a problem in the combination of db-migration between two versions in combination with the force-reload-flag?

---

The only thing that doesn't look normal on my local machine is on the very first boot of a completely new connector image (current main) various messages about DB migration errors about non-existing constraints which are tried to be removed.

![image](https://user-images.githubusercontent.com/75306992/150503101-0f636eee-b7ef-4494-8dcf-5195bff4ed3b.png)

However, this does not cause any further startup problems, except for the messages that the ALTER TABLE statements accordingly run into the void.

**Comment by HeinrichPet** [comment]
Ronja tested if there is a problem with migration and force-reload: 
![migration](https://user-images.githubusercontent.com/59964830/150512439-d8feb018-33c5-4f8a-a2eb-c4cbd8bdc316.jpg)
However, configurations are not migrated during DB Migration: https://international-data-spaces-association.github.io/DataspaceConnector/Deployment/DatabaseMigration

**Comment by HeinrichPet** [comment]
Let's wait until Monday for @pdrd to give us more info.

**Comment by pdrd** [comment]
Thank you very much for all the effort and time you put into this issue!

**TL;DR** The issue can not be reproduced and seems to be a misinterpretation of the unreplaced "@project.version@" in the self-description and unrelated key constraint errors.

I've done multiple tests with different connectors. Two on my local machine with versions 6.4.0, 6.5.0, 6.5.1 and 6.5.2 and with an existing connector with a customized 6.5.2 with log4j 2.17.1, which threw the error originally in december. These versions were tested, because the connector was upgraded between these versions, before the error occured.

The results show, that in most cases there is no key constraint error showing up and the migration works perfectly fine. The only key constraint problems are the ones reproduced by [Tim](https://github.com/International-Data-Spaces-Association/DataspaceConnector/issues/850#issuecomment-1018339179), but these disappear after a restart.

My conclusion is, that we misinterpreted another issue we wanted to fix in the end of december. Our existing connectors all show the template string "@project.version@" as ids:version in their self-description, but don't replace it with the current version. We were trying to fix it by updating the connector to a new version which supported "configuration.force.reload", but during the update the key constraint errors from [Tim](https://github.com/International-Data-Spaces-Association/DataspaceConnector/issues/850#issuecomment-1018339179) arose.
The connector started with the errors, but the @project.version@ still didn't get replaced and to this time we thought this is about the ids:version getting updated in the database correctly and so we assumed, that the migration didn't work correctly because of the given key constraint errors. With the test results i can confirm, that the migration is working and that even the configuration is successfully updated on a new startup.
Sadly the @project.version@ is still not replaced in the self-description, but this seems to be another issue.

Thank you again for your help and sorry, that it was a false alarm.

**Comment by juliapampus** [comment]
Happy to read that it works. Thank you for your detailed answer! I will close this issue then. We can open up a new one for the version number issue.
